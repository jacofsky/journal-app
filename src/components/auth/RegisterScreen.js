import React, { useEffect } from 'react'
import {useDispatch, useSelector} from 'react-redux'

import { Link } from 'react-router-dom';
import validator from 'validator';


import { useForm } from '../../hooks/useForm';
import { setError, removeError } from '../../actions/ui';
import { startRegisterWithEmailPasswordName } from '../../actions/auth';


const RegisterScreen = () => {

    const dispatch = useDispatch()
    const {msgError} = useSelector( state => state.ui )

    useEffect(() => {
        dispatch(removeError())
    }, [])

    const {formValues, handleInputChange} = useForm({
        name: 'Pablin',
        email: 'Pablis@gmail.com',
        password: '123456',
        password2: '123456'
    })

    const {name, email, password, password2} = formValues;

    const handleRegister = (e) => {
        e.preventDefault()

        if(isFormValid()) {
            dispatch(startRegisterWithEmailPasswordName(email, password, name))
        }
    }

    const isFormValid = () => {
        
        if (validator.isEmpty(name)) {
            dispatch(setError('Name is required'))

            return false
        } else if (!validator.isEmail(email)){
            dispatch(setError('Email is invalid'))


            return false 
        } else if (password !== password2 || password.length < 5 || password2.length < 5) {
            dispatch(setError('Password shold be at least 6 caracters and match each other'))

            return false 
        }

        dispatch(removeError())
        return true
    }

    return (
        <div className='animate__animated animate__fadeIn animate__faster'>
            <h3 className='auth__tittle'>Register</h3>
            
            {
                msgError 
                &&
                <div className="auth__alert-error my-2">
                    {msgError}
                </div>
            }
            

            <form onSubmit={handleRegister}>
                <div className="row mb-3">
                    <div>
                        <input className='w-100 mb-2 auth__input' type="text" placeholder='Name' name='name' onChange={handleInputChange} value={name} />
                        <input className='w-100 mb-2 auth__input' type="text" placeholder='Email' name='email' onChange={handleInputChange} value={email} />
                        <input className='w-100 mt-2 auth__input' type="password" placeholder='Password' name='password' onChange={handleInputChange} value={password} />
                        <input className='w-100 mt-2 auth__input' type="password" placeholder='Confirm password' name='password2' onChange={handleInputChange} value={password2} />
                    </div>
                    <div className='justify-content-center d-flex mt-3'>
                        <button className="auth__button btn w-100" type="submit">Register <i className="bi bi-arrow-bar-right"></i></button>
                    </div>
                    
                </div>

                <Link to="/auth/login" className='link'>
                    Already registered?
                </Link>
                
            </form>
        </div>
    )
}

export default RegisterScreen
