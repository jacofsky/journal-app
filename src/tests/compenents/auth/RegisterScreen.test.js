import React from "react"
import { Provider } from "react-redux"
import { MemoryRouter } from "react-router-dom"
import configureStore from 'redux-mock-store'
import thunk from 'redux-thunk'
import { mount } from "enzyme"
import RegisterScreen from "../../../components/auth/RegisterScreen"
import { types } from "../../../types/types"



const middlewares = [thunk]
const mockStore = configureStore(middlewares)

const initState = {

    auth: {},
    ui: {
        loading: false,
        msgError: null
    }
}

let store = mockStore(initState)

const wrapper = mount(
    <Provider store={store}>
        <MemoryRouter>
            <RegisterScreen />
        </MemoryRouter>
    </Provider>
)

describe('Tests in RegisterScreen', () => {


    test('should match snapshot', () => {

        expect(wrapper).toMatchSnapshot()
    
    })

    test('should do dispatch of the respective action', () => { 

        const emailField = wrapper.find('input[name="email"]')

        emailField.simulate('change', {
            target: {
                value: '', 
                name: 'email'
            }
        })

        wrapper.find('form').simulate('submit', {
            preventDefault(){}
        })

        const actions = store.getActions()

        expect(actions[0].type).toBe(types.uiRemoveError)
    })

    test('should show error message', () => { 
        
        const initState = {

            auth: {},
            ui: {
                loading: false,
                msgError: 'Email no es correcto'
            }
        }
        
        const store = mockStore(initState)
        
        const wrapper = mount(
            <Provider store={store}>
                <MemoryRouter>
                    <RegisterScreen />
                </MemoryRouter>
            </Provider>
        )

        expect(wrapper.find(".auth__alert-error").exists()).toBe(true)
        expect(wrapper.find(".auth__alert-error").text().trim()).toBe(initState.ui.msgError)

     })



})