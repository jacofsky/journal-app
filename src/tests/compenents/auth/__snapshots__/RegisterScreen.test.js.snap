// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests in RegisterScreen should match snapshot 1`] = `
<div
  className="animate__animated animate__fadeIn animate__faster"
>
  <h3
    className="auth__tittle"
  >
    Register
  </h3>
  <form
    onSubmit={[Function]}
  >
    <div
      className="row mb-3"
    >
      <div>
        <input
          className="w-100 mb-2 auth__input"
          name="name"
          onChange={[Function]}
          placeholder="Name"
          type="text"
          value="Pablin"
        />
        <input
          className="w-100 mb-2 auth__input"
          name="email"
          onChange={[Function]}
          placeholder="Email"
          type="text"
          value="Pablis@gmail.com"
        />
        <input
          className="w-100 mt-2 auth__input"
          name="password"
          onChange={[Function]}
          placeholder="Password"
          type="password"
          value="123456"
        />
        <input
          className="w-100 mt-2 auth__input"
          name="password2"
          onChange={[Function]}
          placeholder="Confirm password"
          type="password"
          value="123456"
        />
      </div>
      <div
        className="justify-content-center d-flex mt-3"
      >
        <button
          className="auth__button btn w-100"
          type="submit"
        >
          Register 
          <i
            className="bi bi-arrow-bar-right"
          />
        </button>
      </div>
    </div>
    <a
      className="link"
      href="/auth/login"
      onClick={[Function]}
    >
      Already registered?
    </a>
  </form>
</div>
`;
